src/System.Memory/src/System/SpanHelpers.cs

```diff
@@ -0,0 +1,97 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Reflection;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+
+namespace System
+{
+    internal static class SpanHelpers 
+    {
+        /// <summary>
+        /// Computes "start + index * sizeof(T)", using the unsigned IntPtr-sized multiplication for 32 and 64 bits.
+        /// 
+        /// Assumptions:
+        ///     Start and index are non-negative, and already pre-validated to be within the valid range of their containing Span.
+        ///     
+        ///     If the byte length (Span.Length * sizeof(T)) does an unsigned overflow (i.e. the buffer wraps or is too big to fit within the address space),
+        ///     the behavior is undefined.
+        ///     
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static IntPtr Add<T>(this IntPtr start, int index)
+        {
+            Debug.Assert(start.ToInt64() >= 0);
+            Debug.Assert(index >= 0);
+
+            unsafe
+            {
+                if (sizeof(IntPtr) == sizeof(int))
+                {
+                    // 32-bit path.
+                    uint byteLength = (uint)index * (uint)Unsafe.SizeOf<T>();
+                    return (IntPtr)(((byte*)start) + byteLength);
+                }
+                else
+                {
+                    // 64-bit path.
+                    ulong byteLength = (ulong)index * (ulong)Unsafe.SizeOf<T>();
+                    return (IntPtr)(((byte*)start) + byteLength);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Returns a cached empty array (Array.Empty not available on S.R. 4.0.0.0)
+        /// </summary>
+        public static T[] EmptyArray<T>() => PerTypeLatches<T>.EmptyArray;
+
+        /// <summary>
+        /// Determine if a type is eligible for storage in unmanaged memory. TODO: To be replaced by a ContainsReference() api.
+        /// </summary>
+        public static bool IsReferenceFree<T>() => PerTypeLatches<T>.IsReferenceFree;
```

Comment:
It might be worth putting some code like:

```
if (typeof(T) == typeof(int))
```

so that the JIT can optimize away the branches for those known types